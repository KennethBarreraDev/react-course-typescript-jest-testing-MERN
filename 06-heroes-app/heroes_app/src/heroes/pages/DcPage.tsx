import React from "react"
import { HeroList } from "../components/HeroList"
import { AllowedHeroPublisher } from "../data/heroes"

export const DcPage = React.memo(
  () => {
    return (
     <>
     <h1>Dc comics</h1>
     <hr />
     <HeroList publisher={AllowedHeroPublisher.DC_COMICS}/>
     </>
    )
  }
)
