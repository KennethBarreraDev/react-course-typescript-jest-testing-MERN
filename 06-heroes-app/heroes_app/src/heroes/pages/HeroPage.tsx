import React, { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router";
import { getHeroById } from "../helpers/getHeroById";

export const HeroPage = React.memo(() => {
  let { id } = useParams();
  const hero = useMemo(() => getHeroById(id ?? ''), [])
  const navigate = useNavigate()

  const onNavigateBack = () => {
    navigate('/')
  }

  if (!hero) {
    return <Navigate to='/' />
  }


  return (
    <div className="row mt-5">
      <div className="col-4 animate__animated animate__bounce">
        <img src={`/assets/heroes/${hero.id}.jpg`} alt={hero.superhero} className="img-thumbnail" />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alter ego:</b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher:</b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First apperance:</b> {hero.first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button className="btn btn-primary" onClick={() => onNavigateBack()}>
          Back
        </button>
      </div>
    </div>
  )
})
