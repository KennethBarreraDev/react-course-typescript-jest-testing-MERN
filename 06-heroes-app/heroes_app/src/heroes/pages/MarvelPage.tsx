import React from "react"
import { AllowedHeroPublisher } from "../data/heroes"
import { HeroList } from "../components/HeroList"

export const MarvelPage = React.memo(() => {
  return (
    <>
      <h1>
        Marvel comics
      </h1>
      <hr />
       <HeroList publisher={AllowedHeroPublisher.MARVEL_COMICS}/>
    </>
  )
}
)