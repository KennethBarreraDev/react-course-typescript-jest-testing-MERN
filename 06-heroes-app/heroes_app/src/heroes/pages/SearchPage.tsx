import React, { useMemo } from "react";
import { useCustomForm } from "../../globals/hooks/useForm";
import { useNavigate, useSearchParams } from "react-router";
import { getHeroesByName } from "../helpers/getHeroesByName";
import { HeroCard } from "../components/HeroCard";


export const SearchPage = React.memo(() => {
 
 const navigate = useNavigate()
 
 let [searchParams] = useSearchParams();
 const query = searchParams.get("q") ?? '';
 const {formState, onInputChange} = useCustomForm({searchText: query})
 
 const foundHero = useMemo(() => getHeroesByName(query), [query])

 const onSearchSumbit = (event: React.FormEvent<HTMLFormElement>)=>{
  event.preventDefault()
  if((formState?.searchText ?? '').trim()===''){
    return;
  }
  navigate(`/search?q=${formState?.searchText}`)
 }
  return (
    <>
      <h1>Search</h1>
      <hr />
      <div className="row">
        
      <div className="col-5">
        <form onSubmit={(event)=>onSearchSumbit(event)} role="form">
          <input 
          type="text" 
          placeholder="search a hero"
          className="form-control"
          name="searchText"
          autoComplete="off"
          value={formState?.searchText || ''}
          onChange={(event)=>onInputChange(event)}
          />
          <button className="btn btn-primary">
            Search
          </button>
        </form>
      </div>
      <div className="col-7">
        <h4>Results</h4>
        <hr />

        {
          !foundHero && <div className="alert alert-danger">
          No results for {query}
        </div>
        }

       {
        foundHero &&  <HeroCard hero={foundHero}/>
       }
      </div>
      </div>
    </>

  )
}
)