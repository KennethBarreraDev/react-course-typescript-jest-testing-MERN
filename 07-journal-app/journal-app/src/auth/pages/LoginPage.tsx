import { Google } from "@mui/icons-material";
import { Alert, Button, Grid2, TextField, Typography } from "@mui/material";
import { Link } from "react-router";
import { AppRoutes } from "../../globals/routes";
import { AuthLayout } from "../layout/AuthLayout";
import { useCustomForm } from "../../globals/hooks/useForm";
import React from "react";
import { startLoginWithEmail, startGoogleLogin } from "../../store/auth/authThunks";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store/store";
import { AuthStatus } from "../../store/auth/authSlice";

const initialFormState = { 'email': '', 'password': '' };
export const LoginPage = React.memo(() => {

  const { formState, onInputChange } = useCustomForm(initialFormState);

  const loginStatus = useSelector((state: RootState) => state.auth.status)
  const errorMessage = useSelector((state: RootState) => state.auth.errorMessage)
  const dispatch = useDispatch();

  const onSubmitForm = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const loginUserThunk = startLoginWithEmail((formState ?? {}).email || '', (formState ?? {}).password ?? '')
    loginUserThunk(dispatch)
  };

  const onGoogleSignIn = () => {
    const startGoogleThunk = startGoogleLogin();
    startGoogleThunk(dispatch)
  }

  return (
    <AuthLayout descriptor="Login">
      <form onSubmit={(event) => onSubmitForm(event)} role="form">
        <Grid2 container>
          <Grid2 size={12}>
            <TextField 
              label="email"
              type="email"
               name="email"
              value={(formState ?? {})['email'] || ''}
              placeholder="mail@mail.com"
              fullWidth
              onChange={(event) => onInputChange(event)}
            />
          </Grid2>

          <Grid2 size={12} sx={{ mt: 2 }}>
            <TextField
              name="password"
              label="password"
              type="password"
              aria-label="password-input"
              value={(formState ?? {})['password'] || ''}
              placeholder="****"
              fullWidth
              onChange={(event) => onInputChange(event)}
            />
          </Grid2>
        </Grid2>

        <Grid2 container spacing={2}>
          <Grid2 size={6} sx={{ mb: 2 }}>
            <Button type="submit" variant="contained" fullWidth>
              Login
            </Button>
          </Grid2>

          <Grid2 size={6} sx={{ mb: 2 }}>
            <Button aria-label="google-sign-in" variant="contained" fullWidth onClick={() => onGoogleSignIn()} disabled={loginStatus == AuthStatus.CHECKING ? true : false}>
              <Google />
              <Typography sx={{ ml: 1 }}>Login</Typography>
            </Button>
          </Grid2>
        </Grid2>


        <Grid2 container direction='row' justifyContent='end'>
          <Link to={`/${AppRoutes.auth}/${AppRoutes.register}`}>
            Create an account
          </Link>
        </Grid2>


        {
          (errorMessage != null) &&
          <Grid2 container>
            <Grid2 size={12}>
              <Alert severity="error">{errorMessage ?? ''}</Alert>
            </Grid2>
          </Grid2>
        }
      </form>

    </AuthLayout>
  );
}
)