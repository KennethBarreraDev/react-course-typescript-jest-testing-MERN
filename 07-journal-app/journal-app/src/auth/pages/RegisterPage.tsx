import { Alert, Button, Grid2, TextField } from "@mui/material";
import { AuthLayout } from "../layout/AuthLayout";
import { useCustomForm } from "../../globals/hooks/useForm";
import React from "react";
import { startUserRegister } from "../../store/auth/authThunks";
import { useDispatch, useSelector } from "react-redux";
import { RootState } from "../../store/store";
import { AuthStatus } from "../../store/auth/authSlice";


const initialFormState = { 'fullName': '', 'email': '', 'password': '' };
export const RegisterPage = React.memo(() => {


  const formValidations: { [key: string]: [(value: string) => boolean, string] } = {
    email: [(value: string) => value.includes('@'), 'Please, write a valid email'],
    fullName: [(value: string) => value.length > 0, 'Please, write a valid name'],
    password: [(value: string) => value.length > 6, 'Please, write a valid password']
  }

  const { formState, onInputChange, errors } = useCustomForm(initialFormState, formValidations);

  const dispatch = useDispatch();
  const loginStatus = useSelector((state: RootState) => state.auth.status)
  const errorMessage = useSelector((state: RootState) => state.auth.errorMessage)


  const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const userThunk = startUserRegister(formState!.email, formState!.password, formState!.fullName);
    userThunk(dispatch)
  }

  return (
    <AuthLayout descriptor="Sign in">
      <form onSubmit={(event) => onSubmit(event)} role="form">

        <Grid2 container>
          <Grid2 size={12}>
            <TextField
              error={((errors ?? {})['fullName'] != undefined)}
              helperText={((errors ?? {})['fullName'] || '')}
              label="Full name"
              type="text"
              name="fullName"
              placeholder="Jhon doe"
              value={(formState ?? {})['fullName'] || ''}
              onChange={(event) => onInputChange(event)}

              fullWidth
            />
          </Grid2>
          <Grid2 size={12} sx={{ mt: 2 }}>
            <TextField
              error={((errors ?? {})['email'] != undefined)}
              helperText={((errors ?? {})['email'] || '')}
              label="email"
              type="email"
              name="email"
              placeholder="mail@mail.com"
              value={(formState ?? {})['email'] || ''}
              onChange={(event) => onInputChange(event)}
              fullWidth
            />
          </Grid2>

          <Grid2 size={12} sx={{ mt: 2 }}>
            <TextField
              error={((errors ?? {})['password'] != undefined)}
              helperText={((errors ?? {})['password'] || '')}
              label="password"
              type="password"
              name="password"
              placeholder="****"
              value={(formState ?? {})['password'] || ''}
              onChange={(event) => onInputChange(event)}

              fullWidth
            />
          </Grid2>
        </Grid2>
        <Grid2 container spacing={2} sx={{ mt: 2 }}>
          <Grid2 size={12} sx={{ mb: 2 }}>
            <Button type="submit" variant="contained" fullWidth disabled={loginStatus==AuthStatus.CHECKING ? true : false}>
              register
            </Button>
          </Grid2>
        </Grid2>
        {
          (errorMessage!=null) &&
          <Grid2 container>
          <Grid2 size={12}>
            <Alert severity="error">{errorMessage ?? ''}</Alert>
          </Grid2>
        </Grid2>
        }
      </form>

    </AuthLayout>
  )
})
