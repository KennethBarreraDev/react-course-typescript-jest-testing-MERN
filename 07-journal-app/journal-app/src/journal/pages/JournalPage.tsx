import { IconButton } from "@mui/material"
import { JournalLayout } from "../layout/JournalLayout"
import { NoteView } from "../views/NoteView"
import { AddOutlined } from "@mui/icons-material"
import { useDispatch, useSelector } from "react-redux"
import { startNewNote } from "../../store/journal/journalThunks"
import { RootState } from "../../store/store"
import { NothingSelectedView } from "../views/NothingSelectedView"

export const JournalPage = () => {
  const isSavingNote = useSelector((state: RootState) => state.journal.isSaving)

  const dispatch = useDispatch();

  const onAddNewNote = async () => {
    const startNewNooteThunk = startNewNote();
    startNewNooteThunk(dispatch)

  }

  const activeNote = useSelector((state: RootState) => state.journal.active);

  return (
    <JournalLayout>
      {/* <NothingSelectedView /> */}

      {
        activeNote ? 
        <NoteView /> :
        <NothingSelectedView/>
      }

      <IconButton
        disabled={isSavingNote}
        onClick={() => onAddNewNote()}
        size="large"
        sx={{
          color: 'white',
          backgroundColor:
            'error.main',
          ':hover': 'error.main',
          opacity: 0.9,
          position: "fixed",
          right: 50,
          bottom: 50,
        }}
      >
        <AddOutlined sx={{ fontSize: 30 }} />
      </IconButton>
    </JournalLayout>
  )
}
