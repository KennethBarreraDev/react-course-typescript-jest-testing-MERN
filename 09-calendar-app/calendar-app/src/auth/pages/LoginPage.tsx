import { useEffect } from 'react';
import { useAuthStore } from '../../globals/hooks/useAuthStore';
import { useCustomForm } from '../../globals/hooks/useCustomForm';
import { USER_STATUS } from '../../store/auth/authSlice';
import '../../styles/login.css';
import Swal from "sweetalert2";

const loginFormFields = {
    loginEmail: '',
    loginPassword: '',
}

const registerFormFields = {
    userName: '',
    registerEmail: '',
    registerPassword: '',
    registerPassword2: '',
}

const allFormFields = {
    ...loginFormFields,
    ...registerFormFields
}

export const LoginPage = () => {
    const {
        errorMessage,
        successMesage,
        startCreatingUser,
        status,
        startLogin,

    } = useAuthStore()
    const { formState, onInputChange } = useCustomForm(allFormFields);

    const loginSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        startLogin({ email: formState?.loginEmail || '', password: formState?.loginPassword || '' })
    }

    const onRegisterSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
        startCreatingUser({
            name: formState?.userName || '',
            email: formState?.registerEmail || '',
            password: formState?.registerPassword || ''
        })
    }

    useEffect(() => {
        if (errorMessage.length > 0) {
            Swal.fire('Auth error', errorMessage, 'error')
        }
    }, [errorMessage])

    useEffect(() => {
        if (successMesage.length > 0) {
            Swal.fire('Success creating user', errorMessage, 'success')
        }
    }, [successMesage])

    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-1">
                    <h3>Sign in</h3>
                    <form onSubmit={(event) => loginSubmit(event)}>
                        <div className="form-group mb-2">
                            <input
                                aria-label='loginEmail'
                                type="text"
                                className="form-control"
                                placeholder="Email"
                                name='loginEmail'
                                value={formState?.loginEmail || ''}
                                onChange={(event) => onInputChange(event)}
                            />
                        </div>
                        <div className="form-group mb-2">
                            <input
                                type="password"
                                aria-label='loginPassword'
                                className="form-control"
                                placeholder="Password"
                                name='loginPassword'
                                value={formState?.loginPassword || ''}
                                onChange={status == USER_STATUS.CHECKING ? undefined : (event) => onInputChange(event)}
                            />
                        </div>
                        <div className="form-group mb-2">
                            <input
                                aria-label='login-button'
                                type="submit"
                                className="btnSubmit"
                                value={status == USER_STATUS.CHECKING ? "Loading..." : 'Login'}
                            />
                        </div>
                    </form>
                </div>

                <div className="col-md-6 login-form-2">
                    <h3>Sign up</h3>
                    <form onSubmit={(event) => onRegisterSubmit(event)}>
                        <div className="form-group mb-2">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Full name"
                                name='userName'
                                aria-label='userName'
                                value={formState?.userName || ''}
                                onChange={(event) => onInputChange(event)}
                            />
                        </div>
                        <div className="form-group mb-2">
                            <input
                                type="email"
                                className="form-control"
                                placeholder="Email"
                                aria-label='registerEmail'
                                name='registerEmail'
                                value={formState?.registerEmail || ''}
                                onChange={(event) => onInputChange(event)}
                            />
                        </div>
                        <div className="form-group mb-2">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Password"
                                name='registerPassword'
                                aria-label='registerPassword'
                                value={formState?.registerPassword || ''}
                                onChange={(event) => onInputChange(event)}
                            />
                        </div>

                        <div className="form-group mb-2">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Repeat password"
                                name='registerPassword2'
                                aria-label='registerPassword2'
                                value={formState?.registerPassword2 || ''}
                                onChange={(event) => onInputChange(event)}
                            />
                        </div>

                        <div className="form-group mb-2">
                            <input
                                aria-label='register-button'
                                type="submit"
                                className="btnSubmit"
                                value={status == USER_STATUS.CHECKING ? "Loading..." : 'Create an account'} />
                        </div>
                    </form>
                </div>
            </div>
        </div>
    )
}