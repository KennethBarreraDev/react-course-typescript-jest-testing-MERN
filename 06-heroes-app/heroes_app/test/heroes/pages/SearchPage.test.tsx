import { SearchPage } from "@/heroes/pages/SearchPage"
import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router"


jest.mock('react-router', () => (
  {
    ...jest.requireActual('react-router'),
    useNavigate: jest.fn()
  }
))

describe('SearchPage.test.tsx', () => {
  const navigateFunction = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks();
    (useNavigate as jest.Mock).mockReturnValue(navigateFunction)
  })

  test('Should show hero in screen and input', () => {
    const heroName = 'Batman';
    render(<MemoryRouter initialEntries={[`/search?q=${heroName}`]}>
      <SearchPage />
    </MemoryRouter>)
    expect(screen.getByRole('heading', { level: 5 }).innerHTML).toContain(heroName)
    expect((screen.getByRole('textbox') as HTMLInputElement).value).toBe(heroName)
  })


  test('Should show an error if hero is not found', () => {
    const heroName = 'Gold buster';
    render(<MemoryRouter initialEntries={[`/search?q=${heroName}`]}>
      <SearchPage />
    </MemoryRouter>)
    expect(screen.getByText(`No results for ${heroName}`)).toBeInTheDocument()
  })

  test('Should navigate to hero search', () => {
    const heroName = 'Gold buster';
    render(<MemoryRouter initialEntries={[`/search?q=${heroName}`]}>
      <SearchPage />
    </MemoryRouter>)
    
    const inputElement = screen.getByRole('textbox');
    fireEvent.change(inputElement, { target: { value: heroName } }) 

    const formElement = screen.getByRole('form');
    fireEvent.submit(formElement)

    expect(navigateFunction).toHaveBeenCalled()
  })


})
