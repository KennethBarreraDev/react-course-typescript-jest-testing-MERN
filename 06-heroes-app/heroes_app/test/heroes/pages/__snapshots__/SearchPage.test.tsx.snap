// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage.test.tsx Should show correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="col"
      >
        <div
          class="card"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4"
            >
              <img
                alt="Batman"
                class="card-img"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Batman
                </h5>
                <div>
                  <p>
                    Bruce Wayne
                  </p>
                </div>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  data-discover="true"
                  href="/hero/dc-batman"
                >
                  More...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
